### DISCUSSION TITLE
How to have System.Text.Json skip properties of types that are not serializable?

### DISCUSSION THREAD

**Discussion by WenningQiu** [dotnet/runtime#97473] [General]
I am running into an exception when serializing objects with a property of type `Func<string>`. Since `Func<string>` is code,  it does not make any sense for it to be serialized at all. 
 
```
        public Func<string> MsoCalculator
        {
            get => _msoCalculator ?? GlobalMsoCalculator;
            set => _msoCalculator = value;
        }
```
I was expecting such properties to be automatically excluded by `System.Text.Json.JsonSerializer`, but apparently `System.Text.Json.JsonSerializer` pressed ahead and ran into an exception: 

```
System.NotSupportedException : Serialization and deserialization of 'System.Func`1[[System.String, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]' instances is not supported. The unsupported member type is located on type 'System.Func`1[System.String]'. Path: $.DiagnosticsContext.Mso.MsoCalculator.
  ----> System.NotSupportedException : Serialization and deserialization of 'System.Func`1[[System.String, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]' instances is not supported.
   at System.Text.Json.ThrowHelper.ThrowNotSupportedException(WriteStack& state, NotSupportedException ex)
   at System.Text.Json.Serialization.JsonConverter`1.WriteCore(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo`1.Serialize(Utf8JsonWriter writer, T& rootValue, Object rootValueBoxed)
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo`1.SerializeAsObject(Utf8JsonWriter writer, Object rootValue)
   at System.Text.Json.JsonSerializer.WriteStringAsObject(Object value, JsonTypeInfo jsonTypeInfo)
```

So my questions:

1) Should `System.Text.Json.JsonSerializer` be designed to skip properties of types that are not serializable?
2) Since `System.Text.Json.JsonSerializer` currently does not skip such properties, what is the best way to customize `System.Text.Json.JsonSerializer` to do that? (I know slapping `[JsonIgnore]` attribute on the property will help, but I don't want to do it manually on individual properties.

**Comment by eiriktsarpalis** [comment]
`JsonIgnore` is the recommended way to do this. You can also remove properties programmatically by manipulating the `JsonTypeInfo.Properties` property via the [contract customization APIs](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/custom-contracts).
